### ISSUE TITLE
Enhance CodeLens to show defs that are private and external to the namespace.

### ISSUE THREAD

**Issue by dharrigan** [clojure-lsp/clojure-lsp#508]
**Is your feature request related to a problem? Please describe.**

It's often quite difficult, in large namespaces, to know which functions (for example) are being used outside of that namespace, in other namespaces. Right now, with the help of clj-kondo, clojure-lsp would show me (as a codeLens feature), something like this:

![image](https://user-images.githubusercontent.com/427150/128516755-e4c7133a-55d1-47bb-9c5f-e0f27db0e5bd.png)

What is not clear to me, is whether that function is being used only within the scope of the namespace, or is really being used elsewhere (maybe both, once locally and once externally). Here is another example:

![image](https://user-images.githubusercontent.com/427150/128516935-322d9720-1328-433f-9669-41db6c2efdf2.png)

There are 9 references to that function. This function is only ever used in that namespace. The quickest way (at the moment), is to highlight references, i.e,

![image](https://user-images.githubusercontent.com/427150/128517381-0ec0ba33-c366-41c5-9b20-3548621ff4bc.png)

It would be far nicer if instead codeLens could show me as I'm browsing through the file.

**Describe the solution you'd like**

I think a nice addition to codeLens would be the ability to show whether any defs/defns are private to the namespace and thus allow for a quick and easy way to mark the defs/defns private (maybe a further enhancement migth be for a codeAction to do it all for me - but for now, doing that manually is okay).

It might do something like:

`(defn add-ontology-entity-name > 9 private references)`

or, in the case of a mix:

`(defn add-ontology-entity-name > 6 private references, 3 external references)`

or, in the case of them all being external:

`(defn add-ontology-entity-name > 9 external references)`

Whether or not the wording should change to be something other than `private` or `external` is fine with me (e.g., 9 local references; 5 local references, 4 external references etc...)

**Describe alternatives you've considered**

I can get the information from using `(coc-references)`, but it's something I have to do, and not that friendly, in the sense that it would be nicer if the inline reference information could be shown against the function.

Thank you for your consideration.

-=david=-

**Comment by ericdallo** [comment]
Looks a good idea, I just wonder if ` 6 private references, 3 external references` is not a long name, making the code not that clean with that much info?

**Comment by ericdallo** [comment]
Also how thiw would behave when there are tests for this var?
At the momento we show: `1 reference | 1 test`

**Comment by dharrigan** [comment]
I suppose it would ignore tests as being "external". If it's being used in a test, then that is marked as being so, i.e., 

![image](https://user-images.githubusercontent.com/427150/128519286-5d0a2631-89e5-4155-8967-40e29ca681b1.png)

I guess it would be additional information:

`(defn random-uuid > 4 private, 8 external references | 27 tests)`

**Comment by ericdallo** [comment]
Yeah, `4 private, 8 external references | 27 tests` sounds like the minimal to me

**Comment by dharrigan** [comment]
I have something now after doing a bit of playing around and after that, I'm not sure of the merit.

Correct me if I'm wrong, but I don't think it is possible to have a mix of private and public for a reference.

A reference is ether public or private, cannot be both. I can reference a private var in a namespace or I can reference a public var in the same namespace or other namespaces. Thus, having "X private, Y references" doesn't really make sense.

What I've ended up with is:

```
(def ^:private foo 1)      1 private reference

foo
```

```
(def foo 1)     1 public reference

foo
```

If I reference `foo` from another namespace, then I would have

```
foo     2 public references
```

So, not sure now whether I'm gaining anything.

**Comment by ericdallo** [comment]
yes, you are correct, but for this code:
```foo.clj
(ns foo)
(def abc 123)

(+ 1 abc)
```
```bar.clj
(ns bar (:require [foo :as f]))
(+ 2 f/abc)
```

I would expect having something like:
```
(def abc)      1 public reference | 1 private reference
```

That would work, but I think the text is too long, we should try to think in something cleaner/smaller

**Comment by dharrigan** [comment]
Ah, this is where I think I am doing something different.

I notice that you didn't define `^:private` on abc, so it's "public" in-so-far as clj-kondo is determining it. Thus, the `:private` field on the metadata is `public`, thus it would say `2 public references` really.

-=david=-

**Comment by ericdallo** [comment]
I don't think we need to check if the var is private or not, but just check if there is a reference from outside the namespace that we would consider public and everything inside it, private, WDYT?
